pkg/inject/virtualgateway_envoy.go

```diff
@@ -0,0 +1,121 @@
+package inject
+
+import (
+	"encoding/json"
+	appmesh "github.com/aws/aws-app-mesh-controller-for-k8s/apis/appmesh/v1beta2"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"strings"
+)
+
+const envoyImageStub = "injector-envoy-image"
+const envoyVirtualGatewayEnvMap = `
+{
+  "APPMESH_VIRTUAL_NODE_NAME": "mesh/{{ .MeshName }}/virtualGateway/{{ .VirtualGatewayName }}",
+  "APPMESH_PREVIEW": "{{ .Preview }}",
+  "ENVOY_LOG_LEVEL": "{{ .LogLevel }}",
+  "AWS_REGION": "{{ .AWSRegion }}"
+}
+`
+
+type VirtualGatewayEnvoyVariables struct {
+	AWSRegion          string
+	MeshName           string
+	VirtualGatewayName string
+	Preview            string
+	LogLevel           string
+}
+
+type virtualGatwayEnvoyConfig struct {
+	awsRegion    string
+	preview      bool
+	logLevel     string
+	sidecarImage string
+}
+
+// newVirtualGatewayEnvoyConfig constructs new newVirtualGatewayEnvoyConfig
+func newVirtualGatewayEnvoyConfig(mutatorConfig virtualGatwayEnvoyConfig, ms *appmesh.Mesh, vg *appmesh.VirtualGateway) *virtualGatewayEnvoyConfig {
+	return &virtualGatewayEnvoyConfig{
+		ms:            ms,
+		mutatorConfig: mutatorConfig,
+		vg:            vg,
+	}
+}
+
+var _ PodMutator = &virtualGatewayEnvoyConfig{}
+
+// mutator adding a virtualgateway config to envoy pod
+type virtualGatewayEnvoyConfig struct {
+	vg            *appmesh.VirtualGateway
+	ms            *appmesh.Mesh
+	mutatorConfig virtualGatwayEnvoyConfig
+}
+
+func (m *virtualGatewayEnvoyConfig) mutate(pod *corev1.Pod) error {
+	ok, envoyIdx := containsEnvoyContainer(pod)
+	if !ok {
+		return nil
+	}
+
+	variables := m.buildTemplateVariables(pod)
+	envoyEnv, err := renderTemplate("vgenvoy", envoyVirtualGatewayEnvMap, variables)
+	if err != nil {
+		return err
+	}
+
+	newEnvMap := map[string]string{}
+	err = json.Unmarshal([]byte(envoyEnv), &newEnvMap)
+	if err != nil {
+		return err
+	}
+
+	//we override the image to latest Envoy so customers do not have to manually manage
+	// envoy versions and let controller handle consistency versions across the mesh
+	if pod.Spec.Containers[envoyIdx].Image == envoyImageStub || pod.Spec.Containers[envoyIdx].Image == m.mutatorConfig.sidecarImage {
+		pod.Spec.Containers[envoyIdx].Image = m.mutatorConfig.sidecarImage
+	} else {
+		return errors.Errorf("invalid envoy image name for injection %s, expected name: %s", pod.Spec.Containers[envoyIdx].Image, envoyImageStub)
```

Comment:
I didn't think to mention this before (sorry!), but we will want to reserve the possibility that someone is bringing their own Envoy image for the gateway as well. Would it make more sense to just mutate if `name==envoy &&  image==injector-envoy-image`, otherwise allowing the provided configuration?

Reply:
I thought about it. The downside is that it also opens the possibility of customer programming invalid images like not-envoy or envoy-image-with-missing-features or envoy-image-not-longer-supported. If we're okay with the risk, we can open it up. What do you think?

Reply:
Also, in either case, it's definitely a two-way door decision. And we do currently support overriding sidecar image [1]

[1] https://github.com/aws/aws-app-mesh-controller-for-k8s/blob/v1beta2/pkg/inject/config.go#L76

Reply:
Yeah, I think there’s a rare case where the gateway needs to be a custom envoy while the sidecars may be something else. I think most customers will prefer consistency, but allowing them to bring their own image is something we’ve always wanted to provide as an option (even though we may not be able to fully support it).

Reply:
Sounds good. We can allow that. I will push a subsequent PR after this to allow for custom images

Reply:
Added the option to side override https://github.com/aws/aws-app-mesh-controller-for-k8s/pull/266. Added an annotation instead of using a special image to determine the condition